src/server/gc_worker/gc_manager.rs

```diff
@@ -449,12 +453,33 @@ impl<S: GcSafePointProvider, R: RegionInfoProvider + 'static, E: KvEngine> GcMan
         // The following loop iterates all regions whose leader is on this TiKV and does
         // GC on them. At the same time, check whether safe_point is updated
         // periodically. If it's updated, rewinding will happen.
+
+        let (tx, rx) = mpsc::channel();
+        let cb = Box::new(move |res| {
+            // send must be successful.
+            tx.send(res).unwrap();
+        });
+        let mut running_tasks = 0;
         loop {
             self.gc_manager_ctx.check_stopped()?;
             if is_compaction_filter_allowed(&self.cfg_tracker.value(), &self.feature_gate) {
                 return Ok(());
             }
 
+            if running_tasks >= self.max_concurrent_tasks {
+                let res = rx.recv().unwrap();
```

Comment:
Actually, I think, there should be some way to block it on the task sender which is consumed by the GCWorker or GCRunner, in which way the code should be able to be more clear 🤔

Reply:
@afeinberg It serves the purpose. Thanks. The waitgroup seems more straightforward for this.

Reply:
Actually after looking at the wait group today's implementation, I believe using bounded channel should be better.

Reply:
@afeinberg The problem of using mpsc (regardless bounded or not) is that there's only one receiver.  So we cannot have the callback receive from channel (to reduce number) meanwhile sending channel an flag to indicate scheduling a task.  Because callback can run parallelly with multiple receivers.

Reply:
> Actually, I think, there should be some way to block it on the task sender which is consumed by the GCWorker or GCRunner, in which way the code should be able to be more clear 🤔

Updated a version with CondVar + Mutex.  The code is a bit more complex than the previous version, maybe it's easier to understand though.